File: abyss/src/main/java/abyss/plugin/api/plugin/attributes/GlobalAttributes.java
package abyss.plugin.api.plugin.attributes;

import abyss.plugin.api.Debug;

import java.io.*;

public final class GlobalAttributes {

    public static final Attributes GLOBAL = new Attributes();

    private GlobalAttributes() {}

    public static void save() {
        File file = new File("global_attributes.bin");
        if(!file.exists()) {
            try {
                file.createNewFile();
            } catch (IOException e) {
                Debug.printStackTrace("Failed to create global attributes file", e);
            }
        }
        try (FileOutputStream stream = new FileOutputStream(file)) {
            GLOBAL.flushAndWrite(stream);
        } catch (IOException e) {
            Debug.printStackTrace("Failed to open global attributes", e);
        }
    }

    public static void load() {
        File file = new File("global_attributes.bin");
        if(!file.exists()) {
            return;
        }
        try (FileInputStream stream = new FileInputStream(file)) {
            GLOBAL.read(stream);
        } catch (IOException e) {
            Debug.printStackTrace("Failed to open global attributes", e);
        }
    }

}

File: src/main/kotlin/abyss/plugin/api/actions/attributes/GlobalAttributes.kt
package abyss.plugin.api.actions.attributes

import java.io.*

object GlobalAttributes : PluginAttributeSerializer {

    @JvmStatic
    val attributes = PluginAttributes(HashMap(), this)

    @JvmStatic
    private val file = File("global_attributes.bin")

    @JvmStatic
    fun save() {
        if(!file.exists()) {
            file.createNewFile()
        }
        val out = attributes.serialize(attributes)
        file.writeBytes(out.toByteArray())
    }

    @JvmStatic
    fun load() {
        if(!file.exists()) return
        val inputStream = ByteArrayInputStream(file.readBytes())
        attributes.deserialize(attributes, inputStream)
    }

    override fun serialize(attributes: PluginAttributes): ByteArrayOutputStream {
        val formatStream = ByteArrayOutputStream()
        val data = DataOutputStream(formatStream)
        data.writeShort(attributes.size())
        attributes.forEach { key, value ->
            data.writeUTF(key)
            data.writeUTF(value)
        }

        return formatStream
    }

    override fun deserialize(attributes: PluginAttributes, inputStream: ByteArrayInputStream) {
        val data = DataInputStream(inputStream)
        val size = data.readUnsignedShort()
        repeat(size) {
            val key = data.readUTF()
            val value = data.readUTF()
            attributes.put(key, value)
        }
    }

}